Nullable.md

```diff
@@ -0,0 +1,191 @@
+methods will work on objects of this type unless they are defined for both
+`f(x::Some{T})` and `f(x::Void)`. As such, this approach requires consistent
+of specialized lifting machinery via dot-broadcasting syntax. See the later
+section on lifting for details.
+
+Because users will have to work with lifted functions, this approach is
+less convenient, but more general and often safer. It also allows one to
+distinguish between outcomes in cases where a method returns something like
+`Union{Some{Void}, Void}` -- as might happen in a hash table implementation
+that returns `Some` when a key is present and associated with `nothing` as a
+value, but returns `Void` when a key is not present.
+
+### `Union{T, Null{T}}` Pros & Cons
+
+This approach allows us to distinguish `Null{Bool}` from `Null{Int}` in
+cases in which that distinction might be useful. R makes such a distinction,
+although it seems to be done only because it is falls out naturally from an
+implementation that uses sentinel values. Otherwise, it is similar to the
+`Union{T, Void}` case described above, except for one crucial problem.
+
+That problem is essentially the problem we already have with `Nullable{T}`:
+writing "clean" functions requires information about type-inference because of
+a coupling of result types to knowledge about counterfactual outcomes in
+branches.
+
+To see the issue, consider the following possible implementation of
+`map(f, x::Nullable{T})` in our current implementation of `Nullable{T}`:
+
+```
+function map(f, x::Nullable{T})
+    if hasvalue(x)
+        return Nullable(f(value(x)))
+    else
+        return Nullable{return_types(f, T)}()
+    end
+end
+```
+
+The problem is that, when given a null input, the return type can only match
+the counterfactual return type that would have occurred if called on a non-null
+value if the function knows the return type of `f`.
+
+In static languages with arrow types this is not a problem as the return type
+of a function is guaranteed to be well-defined by the semantics of the
+language. But Julia does not provide such guarantees, so type-inference must
+be invoked to determine information about the return type of the counterfactual
+branch.
+
+Note that this problem does not occur for the non-null case because we are
+expected to evaluate `f`. It is only a problem when we need to know the return
+type of `f`, but must not determine this type by evaluation of `f`.
+
+## Step 2:: Define Core Operations on Nullable Types by Lifting
+
+The core semantic issue for nullable types is maximizing code reuse: we want
+to evaluate `f(x::?T)`, but not require that every function be defined for
+both `f(x::T)` and `f(x::?T)` separately as the overwhelming majority of
+functions will return `null` as their output whenever any of their arguments
+are `null`. Implementing this lifting of functions can be solved as follows:
+
+First, we define `broadcast(f, x::Nullable)` as follows:
```

Comment:
Wait, wait, I'm lost here. If we literally replace `Nullable{T}` with `Union{T, Void}` or `Union{T, Null{T}}` then dispatch will handle function calls. `broadcast`, like *every* other function, would receive a concrete type like `T` or `Void` or `Null{T}`. 

Further, if you mean `typealias Nullable{T} Union{T, Void}`, then we would also have `T <: Nullable` for all possible types `T`.

Like `DataArrays` and `NA`, lifting is not necessary. You **do** need a bunch more method definitions for `Void` / `Null{T}` / `Some{T}`.